File: python/02.py
def memoize(func):
    memo = {}
    def helper(k):
        if k not in memo:
            memo[k] = func(k)
        return memo[k]
    return helper

def fib(k):
    if k <= 2:
        return k
    return fib(k-2) + fib(k-1)
fib = memoize(fib)

result = 0
i = 2
value = fib(2)
while value <= 4000000:
    if value%2==0:
        result += value
    i+=1
    value = fib(i)
print(result)

File: c++/02.cc
#include<iostream>
#include<map>
using namespace std;

long long fib(int k){
    static map<int, long long> memo;
    if(k <= 2) return k;
    if(memo.count(k) > 0)
        return memo[k];
    
    long long result = fib(k-2) + fib(k-1);
    memo[k] = result;
    return result;
}

int main(){
    long long result = 0;
    int i = 2;
    long long value = fib(2);
    while(value <= 4000000){
        if(value%2==0){
            result += value;
        }
        i++;
        value = fib(i);
    }
    cout << result << "\n";
}